generator_gomod.go

```diff
@@ -54,6 +77,100 @@ func withGomod(g *SourceGenerator, srcSt, worker llb.State, opts ...llb.Constrai
 	}
 }
 
+func withCredHelper(credHelper llb.State, credHelperPath string) RunOptFunc {
+	return func(ei *llb.ExecInfo) {
+		llb.AddMount(credHelperPath, credHelper, llb.SourcePath(GitCredentialHelperGomod)).SetRunOption(ei)
+	}
+}
+
+func (g *SourceGenerator) mountGitAuthConfig(mountPoint, basename string) llb.RunOption {
+	return RunOptFunc(func(ei *llb.ExecInfo) {
+		if g.Gomod == nil || g.Gomod.Auth == nil {
+			return
+		}
+
+		var b bytes.Buffer
+		enc := gob.NewEncoder(&b)
+		if err := enc.Encode(g.Gomod.Auth); err != nil {
+			panic("cannot marshal dalec spec bytes")
+		}
+
+		st := llb.Scratch().File(llb.Mkfile("/"+basename, 0o644, b.Bytes()))
+		llb.AddMount(mountPoint, st).SetRunOption(ei)
+	})
+}
+
+func (g *SourceGenerator) gitconfigGeneratorScript(scriptPath, configPath string) llb.State {
+	var script bytes.Buffer
+
+	sortedHosts := SortMapKeys(g.Gomod.Auth)
+	if len(sortedHosts) > 0 {
+		fmt.Fprintln(&script, `set -eu`)
+	}
+
+	for _, host := range sortedHosts {
+		if sshConfig := g.Gomod.Auth[host].SSH; sshConfig != nil {
+			username := "git"
+			if sshConfig.Username != "" {
+				username = sshConfig.Username
+			}
+
+			fmt.Fprintf(&script, `git config --global "url.ssh://%[1]s@%[2]s/.insteadOf" https://%[2]s/`, username, host)
+			script.WriteRune('\n')
+			continue
+		}
+
+		fmt.Fprintf(&script, `git config --global credential."https://%s".helper "gomod %s"`, host, configPath)
+		script.WriteRune('\n')
+	}
+
+	fmt.Fprintln(&script, "go mod download")
+	return llb.Scratch().File(llb.Mkfile(scriptPath, 0o755, script.Bytes()))
+}
+
+func (g *SourceGenerator) withGomodSecretsAndSockets() llb.RunOption {
+	return runOptionFunc(func(ei *llb.ExecInfo) {
+		if g.Gomod == nil {
+			return
+		}
+
+		var (
+			setenv = func() {
+				llb.AddEnv(
+					"GIT_SSH_COMMAND",
+					`ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no`,
+				).SetRunOption(ei)
+			}
+
+			noop = func() {}
+		)
+
+		secrets := make(map[string]struct{}, len(g.Gomod.Auth))
+		for _, auth := range g.Gomod.Auth {
+			if auth.Token != "" {
+				secrets[auth.Token] = struct{}{}
+				continue
+			}
+
+			if auth.Header != "" {
+				secrets[auth.Header] = struct{}{}
+				continue
+			}
+
+			if auth.SSH != nil && auth.SSH.ID != "" {
+				llb.AddSSHSocket(llb.SSHID(auth.SSH.ID)).SetRunOption(ei)
+
+				setenv()
+				setenv = noop
+			}
+		}
+
+		for secret := range secrets {
+			llb.AddSecret("/run/secrets/"+secret, llb.SecretID(secret)).SetRunOption(ei)
```

Comment:
What it help simplify things to mount the secret such that we can find it by host name instead of having to encode configs?
e.g. something like `/run/secrets/dalec-git-auth/<host>`

Reply:
I was thinking the same thing earlier today. I'll set the destination path on the secrets instead.

Reply:
We'll need to encode the auth type also (token or header).